sandbox/apparmor/notify/message.go

```diff
@@ -0,0 +1,418 @@
+// described below.
+//
+//	struct apparmor_notif_resp {
+//	  struct apparmor_notif base;
+//	  __s32 error;        /* error if unchanged */
+//	  __u32 allow;
+//	  __u32 deny;
+//	} __attribute__((packed));
+type MsgNotificationResponse struct {
+	MsgNotification
+	// XXX: The embedded MsgNotification also has an Error field, why?
+	Error int32
+	// Allow somehow encodes the allowed operation mask.
+	Allow uint32
+	// Deny somehow encodes the denied operation mask.
+	Deny uint32
+}
+
+// ResponseForRequest returns a response message for a given request.
+func ResponseForRequest(req *MsgNotification) MsgNotificationResponse {
+	return MsgNotificationResponse{
+		MsgNotification: MsgNotification{
+			NotificationType: APPARMOR_NOTIF_RESP,
+			// XXX: should Signalled be copied?
+			Signalled: req.Signalled,
+			// XXX: should Flags be copied?
+			Flags: req.Flags,
+			ID:    req.ID,
+			// XXX: should Error be copied?
+			Error: req.Error,
+		},
+	}
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotificationResponse) MarshalBinary() ([]byte, error) {
+	msg.Version = 2
+	msg.Length = uint16(binary.Size(*msg))
+	buf := bytes.NewBuffer(make([]byte, 0, msg.Length))
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Write(buf, order, msg); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// msgNotificationOpKernel (TBD, document me).
+//
+//	struct apparmor_notif_op {
+//	  struct apparmor_notif base;
+//	  __u32 allow;
+//	  __u32 deny;
+//	  pid_t pid;          /* pid of task causing notification */
+//	  __u32 label;        /* offset into data */
+//	  __u16 class;
+//	  __u16 op;
+//	} __attribute__((packed));
+type msgNotificationOpKernel struct {
+	MsgNotification
+	Allow uint32
+	Deny  uint32
+	Pid   uint32
+	Label uint32
+	Class uint16
+	Op    uint16
+}
+
+// MsgNotificationOp describes a prompt request.
+//
+// The actual information about the prompted object is not encoded here.
+// The mediation class can be used to deduce the type message that was actually sent
+// and decode.
+type MsgNotificationOp struct {
+	MsgNotification
+	// Allow describes the permissions the process, attempting to perform some
+	// an operation, already possessed. Use DecodeFilePermissions to decode it,
+	// if the mediation class is AA_CLASS_FILE.
+	Allow uint32
+	// Deny describes the permissions the process, attempting to perform some
+	// operation, currently lacks. Use DecodeFilePermissions to decode it, if
+	// the mediation class is AA_CLASS_FILE.
+	Deny uint32
+	// Pid of the process triggering the notification.
+	Pid uint32
+	// Label is the apparmor label of the process triggering the notification.
+	Label string
+	// Class of the mediation operation.
+	// Currently only AA_CLASS_FILE is implemented in the kernel.
+	Class MediationClass
+	// XXX: This is unused.
+	Op uint16
```

Comment:
Is `Op` still unused in apparmor?

Reply:
Op can be set but is not guaranteed, on notifications but is ignored on responses. It provides supplemental information about what operation caused the notification. Atm I would consider it debug type info and can be ignored by snapd